hwy/contrib/random/random.h

```diff
@@ -0,0 +1,219 @@
+
+    return result;
+  }
+
+ public:
+
+  /* This is the jump function for the generator. It is equivalent
+ to 2^128 calls to next(); it can be used to generate 2^128
+ non-overlapping subsequences for parallel computations. */
+  constexpr void Jump() noexcept {
+    constexpr std::uint64_t JUMP[] = {0x180ec6d33cfd0aba, 0xd5a61266f0c9392c,
+                                      0xa9582618e03fc9aa, 0x39abdc4529b1661c};
+    std::uint64_t s0 = 0;
+    std::uint64_t s1 = 0;
+    std::uint64_t s2 = 0;
+    std::uint64_t s3 = 0;
+    for (auto i : JUMP)
+      for (auto b = 0; b < 64; b++) {
+        if (i & std::uint64_t{1} << b) {
+          s0 ^= m_state[0];
+          s1 ^= m_state[1];
+          s2 ^= m_state[2];
+          s3 ^= m_state[3];
+        }
+        Next();
+      }
+
+    m_state[0] = s0;
+    m_state[1] = s1;
+    m_state[2] = s2;
+    m_state[3] = s3;
+  }
+
+};
+}  // namespace internal
+template <typename T>
+class HWY_CONTRIB_DLLEXPORT VectorXoshiro {
+ public:
+  explicit VectorXoshiro(const std::uint64_t seed) {
+    namespace hn = hwy::HWY_NAMESPACE;
+    internal::Xoshiro xoshiro{seed};
+    const auto lanes = hn::Lanes(hn::DFromV<T>());
+    auto stateArray = hwy::MakeUniqueAlignedArray<std::uint64_t>(StateSize());
+
+    for (auto i = 0UL; i < lanes; ++i) {
+      const auto state = xoshiro.GetState();
+      for (auto j = 0UL; j < internal::Xoshiro::StateSize(); ++j) {
+        const auto index = lanes * j + i;
+        stateArray[index] = state[j];
+      }
+      xoshiro.Jump();
+    }
+
+    for (auto i = 0UL; i < internal::Xoshiro::StateSize(); ++i) {
+      m_state[i] = hn::Load(hn::DFromV<T>(), &stateArray[i * lanes]);
+    }
+  }
+
+  constexpr auto operator()() { return Next(); }
+
+  static constexpr std::uint64_t StateSize() noexcept {
+    namespace hn = hwy::HWY_NAMESPACE;
+    return hn::Lanes(hn::DFromV<T>()) * internal::Xoshiro::StateSize();
+  }
+
+  std::array<std::uint64_t, StateSize()> GetState() const {
+    namespace hn = hwy::HWY_NAMESPACE;
+    const auto lanes = hn::Lanes(hn::DFromV<T>());
+    auto stateArray = hwy::MakeUniqueAlignedArray<std::uint64_t>(StateSize());
+    for (auto i = 0UL; i < internal::Xoshiro::StateSize(); ++i) {
+      hn::Store(m_state[i], hn::DFromV<T>(), stateArray.get() + i * lanes);
+    }
+    std::array<std::uint64_t, StateSize()> state;
+
+    auto index = 0UL;
+    for (auto i = 0UL; i < lanes; ++i) {
+      for (auto j = 0UL; j < internal::Xoshiro::StateSize(); ++j) {
+        state[index++] = stateArray[lanes * j + i];
+      }
+    }
+    return state;
+  }
+
+  constexpr auto Uniform() noexcept {
+    namespace hn = hwy::HWY_NAMESPACE;
+    const auto bits = hn::ShiftRight<11>(Next());
+    const auto real = hn::ConvertTo(fl, bits);
+    return real * MUL_VALUE;
+  }
+
+ private:
+  T m_state[4];
+
+  static constexpr hwy::HWY_NAMESPACE::ScalableTag<double> fl{};
+
+  const decltype(hwy::HWY_NAMESPACE::Undefined(fl)) MUL_VALUE =
```

Comment:
Perhaps preferable because shorter: hn::Vec<decltype(fl)>?
BTW I'd recommend some kind of prefix to make clear we're dealing with the blueprint/descriptor, rather than an actual vector. Perhaps d_fl or tag_fl?